sql/colexec: TestMergeJoinerMultiBatchRuns failed

sql/colexec.TestMergeJoinerMultiBatchRuns [failed](https://tanzanite.cluster.engflow.com/invocations/default/6287aeea-7c39-42e9-a113-00661c807d4c?testReportRun=14&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvbGV4ZWM6Y29sZXhlY190ZXN0) on master @ [d0582983521533ffc15a8b8525a188d80da2a0b7](https://github.com/cockroachdb/cockroach/commits/d0582983521533ffc15a8b8525a188d80da2a0b7):

Fatal error:

```
panic: test-mem: no bytes in account to release, current 0, free 659 [recovered]
	panic: test-mem: no bytes in account to release, current 0, free 659

```
Stack: 

```
goroutine 321 [running]:
testing.tRunner.func1.2({0x32f0b60, 0xc000b15758})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x32f0b60?, 0xc000b15758?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4562718, 0x6a2f280}, 0xc002508000, {0x361021a?, 0xc001285a38?}, {0xc001285a00?, 0x0?, 0xc0008fb2c0?})
	github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:417 +0x1e7
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc0024db3c8, {0x4562718, 0x6a2f280}, 0x293)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1108 +0x1d6
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:475 +0x13e
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close(0xc0003b72c0, {0x4562718, 0x6a2f280})
	github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:518 +0x487
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Close(0xc0005026e0, {0x4562718, 0x6a2f280})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:544 +0x102
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Reset(0xc0005026e0, {0x4562718?, 0x6a2f280?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:574 +0x26
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoinerBase).Reset(0xc000502840, {0x4562718?, 0x6a2f280?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:479 +0x35
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*mergeJoinInnerOp).Next(0xc001c1a428)
	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecjoin/mergejoiner_inner.eg.go:10974 +0x5f1
github.com/cockroachdb/cockroach/pkg/sql/colexec.TestMergeJoinerMultiBatchRuns.func1(0xc0004b01a0)
	github.com/cockroachdb/cockroach/pkg/sql/colexec/mergejoiner_test.go:1911 +0x649
testing.tRunner(0xc0004b01a0, 0xc0004be900)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 315
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestMergeJoinerMultiBatchRuns
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMergeJoinerMultiBatchRuns2805439539
    test_log_scope.go:81: use -show-logs to present logs inline
--- FAIL: TestMergeJoinerMultiBatchRuns (0.01s)
=== RUN   TestMergeJoinerMultiBatchRuns/groupSize=468/numInputBatches=1
    --- FAIL: TestMergeJoinerMultiBatchRuns/groupSize=468/numInputBatches=1 (0.01s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=14</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeJoinerMultiBatchRuns.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39308

--- a/pkg/sql/colexec/colexecjoin/mergejoiner_test.go
+++ b/pkg/sql/colexec/colexecjoin/mergejoiner_test.go
@@ -60,6 +60,8 @@ func TestMergeJoinCrossProduct(t *testing.T) {
 	nTuples := 2*coldata.BatchSize() + 1
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
+	diskQueueMemAcc := testMemMonitor.MakeBoundAccount()
+	defer diskQueueMemAcc.Close(ctx)
 	rng, _ := randutil.NewTestRand()
 	typs := []*types.T{types.Int, types.Bytes, types.Decimal}
 	colsLeft := make([]*coldata.Vec, len(typs))
@@ -101,7 +103,7 @@ func TestMergeJoinCrossProduct(t *testing.T) {
 		leftMJSource, rightMJSource, typs, typs,
 		[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 		[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
-		testDiskAcc, testMemAcc, evalCtx,
+		testDiskAcc, &diskQueueMemAcc, evalCtx,
 	)
 	mj.Init(ctx)
 	hj := NewHashJoiner(NewHashJoinerArgs{

--- a/pkg/sql/colexec/colexecutils/spilling_buffer.go
+++ b/pkg/sql/colexec/colexecutils/spilling_buffer.go
@@ -107,6 +107,11 @@ func NewSpillingBuffer(
 	diskQueueMemAcc *mon.BoundAccount,
 	colIdxs ...int,
 ) *SpillingBuffer {
+	if unlimitedAllocator.Acc() == diskQueueMemAcc {
+		colexecerror.InternalError(errors.AssertionFailedf(
+			"memory accounts for allocator and disk queue must be different",
+		))
+	}
 	if colIdxs == nil {
 		colIdxs = make([]int, len(inputTypes))
 		for i := range colIdxs {

--- a/pkg/sql/colexec/colexecutils/spilling_queue.go
+++ b/pkg/sql/colexec/colexecutils/spilling_queue.go
@@ -115,6 +115,11 @@ type NewSpillingQueueArgs struct {
 // If fdSemaphore is nil, no Acquire or Release calls will happen. The caller
 // may want to do this if requesting FDs up front.
 func NewSpillingQueue(args *NewSpillingQueueArgs) *SpillingQueue {
+	if args.UnlimitedAllocator.Acc() == args.DiskQueueMemAcc {
+		colexecerror.InternalError(errors.AssertionFailedf(
+			"memory accounts for allocator and disk queue must be different",
+		))
+	}
 	var items []coldata.Batch
 	if args.MemoryLimit > 0 {
 		items = make([]coldata.Batch, spillingQueueInitialItemsLen)

--- a/pkg/sql/colexec/mergejoiner_test.go
+++ b/pkg/sql/colexec/mergejoiner_test.go
@@ -1739,6 +1739,8 @@ func TestFullOuterMergeJoinWithMaximumNumberOfGroups(t *testing.T) {
 	nTuples := coldata.BatchSize() * 4
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
+	diskQueueMemAcc := testMemMonitor.MakeBoundAccount()
+	defer diskQueueMemAcc.Close(ctx)
 	typs := []*types.T{types.Int}
 	colsLeft := []*coldata.Vec{testAllocator.NewVec(typs[0], nTuples)}
 	colsRight := []*coldata.Vec{testAllocator.NewVec(typs[0], nTuples)}
@@ -1756,7 +1758,7 @@ func TestFullOuterMergeJoinWithMaximumNumberOfGroups(t *testing.T) {
 		leftSource, rightSource, typs, typs,
 		[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 		[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
-		testDiskAcc, testMemAcc, &evalCtx,
+		testDiskAcc, &diskQueueMemAcc, &evalCtx,
 	)
 	a.Init(ctx)
 	i, count, expVal := 0, 0, int64(0)
@@ -1808,6 +1810,8 @@ func TestMergeJoinerMultiBatch(t *testing.T) {
 	defer evalCtx.Stop(ctx)
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
+	diskQueueMemAcc := testMemMonitor.MakeBoundAccount()
+	defer diskQueueMemAcc.Close(ctx)
 	for _, numInputBatches := range []int{1, 2, 16} {
 		t.Run(fmt.Sprintf("numInputBatches=%d", numInputBatches),
 			func(t *testing.T) {
@@ -1826,7 +1830,7 @@ func TestMergeJoinerMultiBatch(t *testing.T) {
 					leftSource, rightSource, typs, typs,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
-					testDiskAcc, testMemAcc, &evalCtx,
+					testDiskAcc, &diskQueueMemAcc, &evalCtx,
 				)
 				a.Init(ctx)
 				i := 0
@@ -1864,6 +1868,8 @@ func TestMergeJoinerMultiBatchRuns(t *testing.T) {
 	defer evalCtx.Stop(ctx)
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
+	diskQueueMemAcc := testMemMonitor.MakeBoundAccount()
+	defer diskQueueMemAcc.Close(ctx)
 	for _, groupSize := range []int{coldata.BatchSize() / 8, coldata.BatchSize() / 4, coldata.BatchSize() / 2} {
 		if groupSize == 0 {
 			// We might be varying coldata.BatchSize() so that when it is divided by
@@ -1901,7 +1907,7 @@ func TestMergeJoinerMultiBatchRuns(t *testing.T) {
 						leftSource, rightSource, typs, typs,
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}, {ColIdx: 1, Direction: execinfrapb.Ordering_Column_ASC}},
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}, {ColIdx: 1, Direction: execinfrapb.Ordering_Column_ASC}},
-						testDiskAcc, testMemAcc, &evalCtx,
+						testDiskAcc, &diskQueueMemAcc, &evalCtx,
 					)
 					a.Init(ctx)
 					i := 0
@@ -2007,6 +2013,8 @@ func TestMergeJoinerRandomized(t *testing.T) {
 	defer evalCtx.Stop(ctx)
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
+	diskQueueMemAcc := testMemMonitor.MakeBoundAccount()
+	defer diskQueueMemAcc.Close(ctx)
 	for _, numInputBatches := range []int{1, 2, 16, 256} {
 		for _, maxRunLength := range []int64{2, 3, 100} {
 			for _, skipValues := range []bool{false, true} {
@@ -2024,7 +2032,7 @@ func TestMergeJoinerRandomized(t *testing.T) {
 						leftSource, rightSource, typs, typs,
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
-						testDiskAcc, testMemAcc, &evalCtx,
+						testDiskAcc, &diskQueueMemAcc, &evalCtx,
 					)
 					a.Init(ctx)
 					i := 0

--- a/pkg/sql/colmem/allocator.go
+++ b/pkg/sql/colmem/allocator.go
@@ -502,6 +502,11 @@ func (a *Allocator) Used() int64 {
 	return a.acc.Used()
 }
 
+// Acc returns the memory account of the Allocator.
+func (a *Allocator) Acc() *mon.BoundAccount {
+	return a.acc
+}
+
 // adjustMemoryUsage adjusts the number of bytes currently allocated through
 // this allocator by delta bytes (which can be both positive or negative).
 //

